### ISSUE TITLE
Make getters take pointers to a `const` buffer

### ISSUE THREAD

**Issue by ubitux** [harfbuzz/harfbuzz#2873]
Would it be possible to have a bunch of functions of the public API to take a pointer to `const` buffer, and in particular `hb_buffer_get_length`? Basically:

```diff
diff --git a/src/hb-buffer.cc b/src/hb-buffer.cc
index 0e293f6b5..8a6114c76 100644
--- a/src/hb-buffer.cc
+++ b/src/hb-buffer.cc
@@ -1351,7 +1351,7 @@ hb_buffer_set_length (hb_buffer_t  *buffer,
  * Since: 0.9.2
  **/
 unsigned int
-hb_buffer_get_length (hb_buffer_t *buffer)
+hb_buffer_get_length (const hb_buffer_t *buffer)
 {
   return buffer->len;
 }
diff --git a/src/hb-buffer.h b/src/hb-buffer.h
index 865ccb227..960d04429 100644
--- a/src/hb-buffer.h
+++ b/src/hb-buffer.h
@@ -462,7 +462,7 @@ hb_buffer_set_length (hb_buffer_t  *buffer,
                      unsigned int  length);

 HB_EXTERN unsigned int
-hb_buffer_get_length (hb_buffer_t *buffer);
+hb_buffer_get_length (const hb_buffer_t *buffer);

 /* Getting glyphs out of the buffer */
```

I'm seeing other potential candidates (such as `hb_buffer_get_glyph_infos`) but I'm not familiar enough with the API nor the depreciation process (adding a `const` can be problematic for users with a function pointer to said function) to submit a clean PR.

This function in particular forces me to mark the buffer mutable in relatively large part of the codebase I'm working on, which is kind of problematic.

**Comment by jfkthame** [comment]
While I agree in principle that it would be good to take a `const` pointer here (and probably a bunch of other places), I'm not sure how Behdad will feel about such a change to the public API at this stage.

As a workaround for your codebase, presumably for cases like this where you know the function does not in fact modify anything, you could just create a trivial "wrapper" that takes a `const` pointer and casts away the `const`-ness when calling the real HarfBuzz API. Then you can freely use `const hb_buffer_t*` in your code..

**Comment by ubitux** [comment]
Yeah sure I have workaround where I can lie about the constness; my request is more about what the API contract actually guarantees. I can *assume* it may not mutate my buffer, but I actually don't know. Like, maybe there is some lazy caching system or something going on. And having a given behavior right now doesn't give any guarantee for the future.

**Comment by behdad** [comment]
I'm actually fine doing this. Not doing this better is one of my regrets about the API.

I started doing it may times, but hit a snag with the reference API (eg `hb_face_reference()`).

Go ahead, send a PR and we go from there. I see that most other objects (face, font, set, map) have const for their getters. But not consistently. For example `hb_face_collect_unicodes()` and family don't have `const`. That seems to be an omission. But also `hb-ot-layout.h`, all API there should take a const face as well...

**Comment by behdad** [comment]
I'm also curious to learn about how you are using `hb_buffer_t` in your app as well. Are you keeping them around for long?

**Comment by ubitux** [comment]
> Go ahead, send a PR and we go from there.

I'm really sorry, currently (and likely for the coming months) I won't have time to follow up on an API PR if I make one... I opened that ticket because it was more of an an inconvenience to me than anything, and it looked like something was overlooked (either on harfbuzz side or on mine).

> I'm also curious to learn about how you are using `hb_buffer_t` in your app as well. Are you keeping them around for long?

So, bear with my I'm probably not using it exactly correctly yet, but my current pattern is:
1. split the original text to make different run (current basic layout with linebreaks, possibly bidi or word based later on): this is where the `hb_buffer_t` are created
2. from these `hb_buffer_t`, build an atlas associating glyph codepoint with their rasterized form; here I'm typically in read-only mode on the buffers by using `hb_buffer_get_length`and `hb_buffer_get_glyph_infos`
3. as a 2nd pass, build an alternative representation of the text by using `hb_buffer_get_length`, `hb_buffer_get_glyph_infos`, `hb_buffer_get_glyph_positions`: again here I'm supposedly in read-only mode on the buffers
4. After this point I can just trash everything.

Basically everything that happens between 1 and 4 is read-only (and it can be a quite a lot of things), so I was looking for some guarantees from the API with regards to side effects.

**Comment by behdad** [comment]
Okay no worries. We'll get to it eventually.

Note that we cannot make `hb_buffer_get_glyph_infos` and `hb_buffer_get_glyph_positions` take a const buffer since they return non-const buffer arrays. In Pango previously we added a separated `_readonly()` version of similar API.

**Comment by behdad** [comment]
> * split the original text to make different runs (current basic layout with linebreaks, possibly bidi or word based later on): this is where the `hb_buffer_t` are created and shaped
> * from these `hb_buffer_t`, build an atlas associating all the different glyph codepoint with their rasterized form; here I'm typically in read-only mode on the buffers by using `hb_buffer_get_length`and `hb_buffer_get_glyph_infos`
> * as a 2nd pass, build an alternative representation of the text by using `hb_buffer_get_length`, `hb_buffer_get_glyph_infos`, `hb_buffer_get_glyph_positions`: again here I'm supposedly in read-only mode on the buffers
> * After this point I can just trash everything harfbuzz related.

Thanks. That's what I wanted to know.

An `hb_buffer_t` takes 40 bytes per glyph. That's definitely more than what it takes to record the shaping results for most clients. So we normally expect the buffer to just be populated, shaped, and results taken out of. In such scenario the constness should not get in your way.
